Display a table of contents in the sidebar of man pages

jvns opened the pull request:
Some of git's man pages are very long and have many sections, and I think it would be easier to navigate them on the web if there were a table of contents. For comparison, [man7.org includes a table of contents at the top for its man pages](https://www.man7.org/linux/man-pages/man3/termios.3.html). 

This updates `script/update-docs.rb` to add the data needed to generate the table of contents to the table contents and renders that data in the sidebar.

Here's what the results look right now in context:

<img width="982" height="700" alt="image" src="https://github.com/user-attachments/assets/14668ecc-15f0-495b-9cc3-95470e9123b4" />

Some possible issues that come to mind:

1. It might be confusing to put the table of contents next to the site's overall navigation. It could be better to put the TOC at the top, like man7.org does it.
2. In the screenshot, you can see that it's hard to read long titles that are split across two lines (like `RECOVERING FROM UPSTREAM REBASE`)
3. If putting the table of contents in the sidebar does make sense, then I think it would be nice to set `position: sticky` so that it stays visible as you scroll the page. I ran into some CSS problems getting that to work but would be happy to figure that out.

Also, I tried regenerating all of the generated HTML pages for all git tags, but ran into some issues so I just generated the v2.50.1 tag for now. Am I meant to regenerate them manually or is it normally handled in CI?

jvns committed:
Display a table of contents in the sidebar of man pages

Adds the table of contents to the frontmatter of the generated HTML
pages, then renders it in the Hugo template.

jvns committed:
Regenerate HTML

dscho commented:
> 1. It might be confusing to put the table of contents next to the site's overall navigation. It could be better to put the TOC at the top, like man7.org does it.

I think that's a very valid point. Personally, I would not have looked for the table of contents of the current manual page in the sidebar. I've tried two things:

## Option 1: moving the TOC next to the Versions, Topics and Language menus

<img width="1218" height="811" alt="image" src="https://github.com/user-attachments/assets/8e748f4f-8e64-4464-b89c-f2a907f69778" />

<details><summary>Diff</summary>

```diff
diff --git a/assets/sass/reference.scss b/assets/sass/reference.scss
index 4873a402c..14d5bc423 100644
--- a/assets/sass/reference.scss
+++ b/assets/sass/reference.scss
@@ -111,6 +111,19 @@ h3.plumbing {
     &#reference-languages-trigger {
       float: right;
     }
+
+    // Ensure the TOC trigger floats right like topics/languages
+    &#reference-toc-trigger {
+      float: right;
+    }
+  }
+
+  // Place the "last updated" span inline next to the left-floating versions trigger
+  > span.light {
+    display: inline-block;
+    clear: none;
+    margin-left: 6px;
+    vertical-align: middle;
   }
 }
 
@@ -213,6 +226,28 @@ h3.plumbing {
   }
 }
 
+// Position the Table of Contents dropdown under the right-floated TOC trigger
+#toc-dropdown {
+  width: 300px;
+  right: 12px;
+  padding: 12px;
+  font-weight: normal;
+  line-height: 1;
+
+  ul {
+    margin: 0;
+  }
+
+  li {
+    line-height: $base-line-height * 0.8;
+  }
+
+  a {
+    font-weight: normal;
+    padding: 0;
+  }
+}
+
 ol.reference-previous-versions {
   @extend .unstyled !optional;
   margin-top: 3px;
diff --git a/layouts/_default/baseof.html b/layouts/_default/baseof.html
index e83e4554b..3bbc0c824 100644
--- a/layouts/_default/baseof.html
+++ b/layouts/_default/baseof.html
@@ -123,6 +123,7 @@
           <div id='reference-version'>
             {{ partial "ref/languages.html" . }}
             {{ partial "ref/topics.html" . }}
+            {{ partial "ref/toc.html" . }}
             {{ partial "ref/versions.html" . }}
           </div>
 
diff --git a/layouts/partials/ref/toc.html b/layouts/partials/ref/toc.html
new file mode 100644
index 000000000..d9d21daf4
--- /dev/null
+++ b/layouts/partials/ref/toc.html
@@ -0,0 +1,13 @@
+{{ $headings := .Params.headings }}
+{{ if $headings }}
+<a class="dropdown-trigger" id="reference-toc-trigger" data-panel-id="toc-dropdown" href="#">Table of Contents Γû╛</a>
+<div class='dropdown-panel right' id='toc-dropdown'>
+  <div>
+    <ul class="toc">
+      {{ range $i, $item := $headings }}
+        <li><a href="#{{ .id }}">{{ .text }}</a></li>
+      {{ end }}
+    </ul>
+  </div>
+</div>
+{{ end }}
diff --git a/layouts/partials/sidebar.html b/layouts/partials/sidebar.html
index 8347b1d2f..69e7c2e72 100644
--- a/layouts/partials/sidebar.html
+++ b/layouts/partials/sidebar.html
@@ -50,19 +50,6 @@
       </li>
     </ul>
 
-    {{ $headings := .Params.headings }}
-    {{ if $headings }}
-    <ul>
-      <li> Table of Contents
-        <ul class="expanded">
-          {{ range $i, $item := $headings }}
-            <li><a href="#{{ .id }}">{{ .text }}</a> </li>
-          {{ end }}
-        </ul>
-      </li>
-    </ul>
-    {{ end }}
-
     {{ if (eq .Params.Sidebar "book") }}
       <hr class="sidebar">
       {{- /* If this page displays a section of the ProGit book, map all the translations thereof */ -}}
```

</details>

## Option 2: Inside the same &lt;div&gt; as the manual page, in the top-right

<img width="1241" height="882" alt="image" src="https://github.com/user-attachments/assets/b29e591a-1a94-4d59-80f4-f08956aef4d4" />

This took a bit of finagling until the text flowed around the table of contents.

<details><summary>Diff</summary>

```diff
diff --git a/assets/sass/reference.scss b/assets/sass/reference.scss
index 4873a402c..383a24dc6 100644
--- a/assets/sass/reference.scss
+++ b/assets/sass/reference.scss
@@ -213,6 +213,68 @@ h3.plumbing {
   }
 }
 
+// New inline TOC box styles
+.toc-box {
+  float: right;
+  width: 200px;
+  margin: 0 0 15px 20px;
+  padding: 10px 12px;
+  font-size: 13px;
+  background-color: var(--callout-color);
+  border: 1px solid var(--base-border-color);
+  @include border-radius(3px);
+
+  .toc-box-header {
+    font-weight: bold;
+    margin-bottom: 8px;
+  }
+
+  .toc-list {
+    margin: 0;
+    padding-left: 18px;
+    list-style: disc;
+  }
+
+  .toc-list li {
+    margin: 2px 0;
+    line-height: $base-line-height * 0.8;
+  }
+
+  .toc-list a {
+    font-weight: normal;
+  }
+}
+
+// Ensure content can flow around the TOC naturally
+#documentation #main {
+  .sectionbody {
+    overflow: visible;
+  }
+
+  .verseblock {
+    overflow: visible;
+
+    .content {
+      overflow: visible;
+
+      pre {
+        overflow: visible;
+        white-space: pre-wrap;
+      }
+    }
+  }
+}
+
+// Stack TOC box full-width on small screens
+@media (max-width: $mobile-m) {
+  .toc-box {
+    float: none;
+    width: 80%;
+    max-width: none;
+    margin-right: 0;
+  }
+}
+
 ol.reference-previous-versions {
   @extend .unstyled !optional;
   margin-top: 3px;
diff --git a/layouts/_default/baseof.html b/layouts/_default/baseof.html
index e83e4554b..75fdcc6a4 100644
--- a/layouts/_default/baseof.html
+++ b/layouts/_default/baseof.html
@@ -135,6 +135,7 @@
 	    <p>Please note that this information is only relevant to you if you plan on <a href="{{ relURL "community#contributing" }}">contributing to the Git project itself</a>. It is in no shape or form required reading for regular Git users.</p>
 	    </div>
 	    {{ end }}
+            {{ partial "ref/toc.html" . }}
             {{ .Content }}
             {{ $match := findRESubmatch "(?s)>NAME</h2>.*?<p[^>]*>(git-)?([^ ]+)" .Content 1 }}
             {{ if (eq ($match | len) 1) }}
diff --git a/layouts/partials/ref/toc.html b/layouts/partials/ref/toc.html
new file mode 100644
index 000000000..1bab34434
--- /dev/null
+++ b/layouts/partials/ref/toc.html
@@ -0,0 +1,11 @@
+{{ $headings := .Params.headings }}
+{{ if $headings }}
+<div id='toc-box' class='toc-box'>
+  <div class='toc-box-header'>Table of Contents</div>
+  <ul class="toc-list">
+    {{ range $i, $item := $headings }}
+      <li><a href="#{{ .id }}">{{ .text }}</a></li>
+    {{ end }}
+  </ul>
+</div>
+{{ end }}
diff --git a/layouts/partials/sidebar.html b/layouts/partials/sidebar.html
index 8347b1d2f..69e7c2e72 100644
--- a/layouts/partials/sidebar.html
+++ b/layouts/partials/sidebar.html
@@ -50,19 +50,6 @@
       </li>
     </ul>
 
-    {{ $headings := .Params.headings }}
-    {{ if $headings }}
-    <ul>
-      <li> Table of Contents
-        <ul class="expanded">
-          {{ range $i, $item := $headings }}
-            <li><a href="#{{ .id }}">{{ .text }}</a> </li>
-          {{ end }}
-        </ul>
-      </li>
-    </ul>
-    {{ end }}
-
     {{ if (eq .Params.Sidebar "book") }}
       <hr class="sidebar">
       {{- /* If this page displays a section of the ProGit book, map all the translations thereof */ -}}
```

</details>

@jvns what do you think? I like option 2 best, but maybe you want to go for something even different?

> I tried regenerating all of the generated HTML pages for all git tags, but ran into some issues so I just generated the v2.50.1 tag for now. Am I meant to regenerate them manually or is it normally handled in CI?

The idea is to run the [`update-git-version-and-manual-pages.yml`](https://github.com/git/git-scm.com/actions/workflows/update-git-version-and-manual-pages.yml) and [`update-translated-manual-pages.yml`](https://github.com/git/git-scm.com/actions/workflows/update-translated-manual-pages.yml) workflows manually in your own fork, and checking the "re-building" checkbox:

<img width="436" height="311" alt="image" src="https://github.com/user-attachments/assets/47cbef06-0f31-47ac-8a18-fa2e4f33e0a2" />

Commit:
Update translated manual pages

Updated via the `update-translated-manual-pages.yml` GitHub workflow.

jvns commented:
I like the way Option 2 looks, though I haven't figured out how to apply the patch locally so I haven't looked at it on mobile yet.

I did a quick survey of other documentation sites to see how they handle this. Most of them put the table of contents for the current page in the sidebar, as well as having a "fixed sidebar" which is always visible as you scroll through the page. A couple of them (MDN, Kotlin) have a separate right sidebar with the TOC for the current page.

This made me think that the sidebar approach is actually pretty normal, but I think for it to really make sense the sidebar would need to be always visible while scrolling the page. Keeping the sidebar always visible is a pretty big design change to how this site works though, so if you don't think that's realistic Option 2 seems fine.

Here are the sites I looked at in my quick docs design survey:

* [python](https://docs.python.org/3/library/datetime.html#timedelta-objects) (left sidebar)
* [ruby](https://docs.ruby-lang.org/en/master/strftime_formatting_rdoc.html#label-Date+-28Year-2C+Month-2C+Day-29) (left sidebar)
* [rust](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize) (left sidebar)
* [go](https://pkg.go.dev/fmt) (left sidebar)
* [swift](https://developer.apple.com/documentation/swift/regex/) (left sidebar)
* [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#syntax)  (right sidebar)
* [Kotlin](https://kotlinlang.org/docs/sequences.html) (right sidebar)
* [Java](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ArrayList.html#hashCode())  (TOC at top)

I'll rerun the workflows in CI.

dscho commented:
@jvns FWIW I ran those workflows in my fork and deployed the site there: https://dscho.github.io/git-scm.com/docs/git-rebase

Commit:
Update manual pages (manually forced rebuild)

Updated via the `update-git-version-and-manual-pages.yml` GitHub workflow.

jvns committed:
Continue to display sidebar when page scrolls

jvns commented:
Here's what it looks like to have the sidebar always visible (mimicking [docs.python.org](https://docs.python.org/3/library/datetime.html#timedelta-objects)'s approach)

<img width="976" height="687" alt="image" src="https://github.com/user-attachments/assets/fcbc51c9-a3bb-4d09-bc19-211cef6cd6c9" />

dscho commented:
Hmm. I am still not so hot about that approach, but I can see where you're coming from.

What would you think about:

- moving the TOC _above_ the navigation?
- separating the TOC and the navigation via a horizontal bar?

jvns commented:
I think either of those could work!

Another idea I just had is to use your `.toc-box` component on smaller screens, but then for screens that are at least 1350px wide, put the table of contents on the right with `position: fixed`, like this:

<img width="1341" height="626" alt="image" src="https://github.com/user-attachments/assets/46ee54f4-c8f1-444f-953d-b4d1c11634d9" />

What I like about that approach is that you can still view the full table of contents on a smaller screen without expanding the sidebar.

I just added this CSS in the browser inspector at https://dscho.github.io/git-scm.com/docs/git-rebase to try it out:

```
.toc-box {
  position: fixed;
  left: calc(940px + 5px + (100vw - 940px)/2);
  margin: 0;
  top: 124px;
  background: #fcfcfa;
  width: 160px;
  max-height: calc(100vh - 150px);
  overflow: scroll;
}
```

dscho commented:
@jvns good idea. I made [the change](https://github.com/dscho/git-scm.com/commit/e96a8276adc47433a85ef36bf0f6dc805e10badc) and am [deploying](https://github.com/dscho/git-scm.com/actions/runs/16883796808) it to my fork.

One thing that still rubs me the wrong way is that when I scroll down, the TOC still stays 124px below the top, and does not align with the top of the `main` &lt;div&gt;... Any ideas?

dscho commented:
@jvns oh, and in Edge, it seems that the scrollbars are always shown, which is a bit unpleasant to my eyes:

<img width="266" height="636" alt="image" src="https://github.com/user-attachments/assets/536b7975-05d4-4c9f-849c-a0667ef0fe87" />

Maybe there is a way to fix that, too?

jvns commented:
> @jvns oh, and in Edge, it seems that the scrollbars are always shown, which is a bit unpleasant to my eyes:

I don't have Edge to test but I think `overflow-y: scroll` would help a bit (so that at least there's no horizontal scrollbar). Also I noticed that the Kotlin docs just completely hide the scrollbars in their table of contents. Not sure how I feel about that

> One thing that still rubs me the wrong way is that when I scroll down, the TOC still stays 124px below the top, and does not align with the top of the main <div>... Any ideas?

I don't know a CSS solution, my only idea for the scrolling thing is to do something like this in Javascript, which I don't love

```
window.addEventListener('scroll', () => {
  const tocBox = document.querySelector('.toc-box');
  const scrollTop = window.pageYOffset || document.documentElement.scrollTop;
  
  tocBox.style.top = scrollTop > 0 ? '0px' : '124px';
});
```

Though one thing the scroll handler makes me think of is that if there were a scroll handler, we could do the thing from [Kotlin docs](https://kotlinlang.org/docs/sequences.html) where the current heading that you're at is highlighted in the sidebar. I think it's a nice effect since it doesn't break the site if someone has JS disabled.

jvns committed:
Differentiate the table of contents a bit more

jvns commented:
Pushed some changes with another idea of how to differentiate the table of contents if it were in the left sidebar 
<img width="808" height="687" alt="image" src="https://github.com/user-attachments/assets/e1d847b4-3128-4c38-8eac-031c776c9d6b" />

dscho commented:
Oooh, nice! What would you think about adding square bullet points, though, in case an entry is multi-line?

jvns committed:
add some square bullet points

jvns commented:
Sure, done! I was hoping to use spacing to make it obvious when something was multiline but I don't think it quite worked.

<img width="199" height="349" alt="image" src="https://github.com/user-attachments/assets/61d1bef4-7208-47a3-878e-afd17bd51c9b" />

dscho commented:
> > One thing that still rubs me the wrong way is that when I scroll down, the TOC still stays 124px below the top, and does not align with the top of the main ... Any ideas?
> 
> I don't know a CSS solution, my only idea for the scrolling thing is to do something like this in Javascript, which I don't love

I finally figured out how to do that: with `position: sticky; float: right` and moving the `<div id="toc-box">` _before_ the `<div id="main">`. Just to prove out the idea, I pushed it to my fork, see e.g. https://dscho.github.io/git-scm.com/docs/git-rebase

But I do tend to like your solution better because it does not _add_ another space but instead reuses the existing navigation bar.

dscho commented:
@jvns so yeah, I like your current version _a lot_. If I would change one thing, it'd be `#toc { margin-right: 0.5rem; }`. What do you think?

| as-is | `margin-right: 0.5rem;` |
| - | - |
|  <img width="1172" height="803" alt="image" src="https://github.com/user-attachments/assets/ed55947f-3445-4984-943d-39a18f5867b6" /> |  <img width="1174" height="811" alt="image" src="https://github.com/user-attachments/assets/40e6c997-8532-4e94-88c9-f9da0d2c5ce1" /> |

jvns committed:
add a margin to the TOC

jvns commented:
Adding a margin sounds good! Just pushed that change. I agree I also like reusing the existing nav bar.

That solution to moving the right navigation is super interesting: I've never learned `float` because I always use something like flexbox or grid instead and I didn't know you could use `float` to do something like that.

dscho reviewed (approved):
Really cool. Thank you so much, once again, for all the great things you're doing Julia!

jvns commented:
Thanks so much for collaborating on it with me (and for everything you do!), it ended up in a much better place than it started :)

dscho commented:
... and 🥁 it's alive! See e.g. https://git-scm.com/docs/git-rebase

kinduff commented:
Glad you landed with the left sidebar, great addition!

To1ne commented:
Hey @jvns, I'm sorry for dropping in late. @dscho pinged me about this ongoing PR, but didn't have time yet to dive into it.

I'm really happy with the solution you've ended up. Putting it in the left sidebar makes way better use of the remaining real estate than the other proposals made in this PR.

Personally I think the CSS can use a little bit of tweaking to make it a bit more pretty, but that's merely some pixel pushing. I'll have a look if I can work on this myself a bit.
